### ISSUE TITLE
Screamer - fatal memory error: Corrupt MCB chain

### ISSUE THREAD

**Issue by superblobster** [joncampbell123/dosbox-x#3939]
### Describe the bug

Screamer crashes frequently with this error.
It happens most consistently on the 4th track of the championship, the one with the port.
I would attach a dump or log, but I cannot find it.
Will test in dosbox 0.74-3 to see if it happens there.
Edit: confirmed, happens there too. Folks on the Steam and GOG forums are also complaining that their Dosbox-bundled copy keeps crashing.

![Screenshot 2023-01-17 080537](https://user-images.githubusercontent.com/60817163/212823502-5dbd2494-ed79-4f4d-9d78-636b0f7b3c23.png)

### Steps to reproduce the behaviour

Play the 4th track of the championship.

### What operating system(s) this bug have occurred on?

Windows 10

### What version(s) of DOSBox-X have this bug?

2022.12.26

### Have you checked that no similar bug report(s) exist?

- [X] I have searched and didn't find any similar bug report.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by joncampbell123** [comment]
Can you try bumping available memory to 32MB, or playing the game with normal core instead of dynamic?

I just purchased this game off of GOG for my testing as well, the default settings use cycles=max and dynamic core.

**Comment by joncampbell123** [comment]
What I'm saying is that whenever a DOS game is unstable or corrupting memory, always try using normal core first.

**Comment by MX9000** [comment]
@joncampbell123 @superblobster 
The track that runs near the port is the fifth and contains a game bug that occurred in SVGA mode even on the hardware of the time: https://www.vogons.org/viewtopic.php?t=19374
The solution is to configure the game in standard VGA mode right before running it.
After completion, you can reconfigure the SVGA resolution.

**Comment by kcgen** [comment]
@joncampbell123 - there's a handy save game here (credit: dosgamert) and repro steps: https://github.com/exoscoriae/eXoDOS/issues/4197#issuecomment-1377472138

Something about track 5's resources (even the low resolution textures) exposes or causes memory mangement bugs in the game. Very interested if you can find fixes or work-arounds for it! 

It's otherwise a great game.

**Comment by joncampbell123** [comment]
I see. It's unusual for a 32-bit protected mode game to have bugs that corrupt conventional memory since most standard malloc() allocation tends to happen in extended memory. Most 32-bit games use conventional memory at most for sound DMA buffers because if EMM386.EXE is loaded, writing memory addresses to DMA is most reliable if in conventional memory.

**Comment by joncampbell123** [comment]
I suppose that since the game is at fault, one possible hack would be to alter INT 21h memory allocation calls to set the memory block some amount larger than what the game requested as a buffer zone.

Another thing you might try are the "debug" options in dosbox.conf that instruct the INT 21h memory allocation to zero the new memory block before returning it to the DOS program. Maybe there's something about the data file and leftover bytes around it that causes the corruption.

Since this is not DOSBox-X's fault, I might close this issue. Closed issues can be browsed easily from GitHub's interface.

**Comment by kcgen** [comment]
> I suppose that since the game is at fault, one possible hack would be to alter INT 21h memory allocation calls to set the memory block some amount larger than what the game requested as a buffer zone.
> 
> Another thing you might try are the "debug" options in dosbox.conf that instruct the INT 21h memory allocation to zero the new memory block before returning it to the DOS program. Maybe there's something about the data file and leftover bytes around it that causes the corruption.

Will poke around with these - great suggestions, @joncampbell123 - thank you.

**Comment by superblobster** [comment]
> one possible hack would be to alter INT 21h memory allocation calls to set the memory block some amount larger than what the game requested as a buffer zone.
> 
> Another thing you might try are the "debug" options in dosbox.conf that instruct the INT 21h memory allocation to zero the new memory block before returning it to the DOS program.

I'm unsure how to do this, but if anybody has tips I'd be happy to try anything.

Increasing available memory doesn't help and the game becomes unplayably laggy in normal core mode.
The crashes seem to happen only when the game runs in SVGA.

Since all the evidence shows that this is an original game bug, I'll close the topic. Perhaps someone will find a workaround one day.

**Comment by grapeli** [comment]
I completed this stage ten times today at SVGA resolution. Five times under dosbox-x and five times under dosbox-staging. Changing the dosbox configuration does not seem to have any effect on trying to bypass this **game's bug**.

The only effective way, apart from changing the resolution to VGA, is to properly position the car on the road at three critical moments.
1. First checkpoint (containers). Always take this straight on the right.

![0001](https://user-images.githubusercontent.com/452325/213194353-8fd94f74-b96b-4eaf-b2f6-f3a64ae032d2.png)

2. Second checkpoint (plane). Always take this straight on the right.

![0002](https://user-images.githubusercontent.com/452325/213194421-45c603d6-a07e-46ed-a56d-e6e96cf4d23f.png)

3. The last skyscraper on the left just before the turn to the start-finish straight. Always drive on the left. 

![0003](https://user-images.githubusercontent.com/452325/213194491-b7cc005b-8a6c-4dcb-8599-2e1d31f9c0e7.png)

Another possible way to get around this error is to change the view. With the view available by pressing F1, this error seems to occur even less frequently. I tested this variant poorly because it is much more difficult to drive the car from this position.

**Comment by kcgen** [comment]
Amazing work, discovering this @grapeli !

**Comment by grapeli** [comment]
I would suggest a little change. People who want to complete this route in SVGA resolution without crashing should change the view to the one available under `F1` or `F2` (`F3` is the default). This will ensure you avoid this game bug.

Choosing the right driving path requires good orientation, prior familiarization with the route and the location of these three critical points.

**Comment by MX9000** [comment]
Good to know @grapeli , thanks for your effort!

**Comment by joncampbell123** [comment]
So would that make the bug a kind of 3D rasterization overflow issue then? Like maybe it computes vertices and triangles after visibility clipping into a fixed sized buffer and that's what the corruption is caused by?

Also the SVGA screenshots definitely show the game's use of affine texture mapping.

**Comment by grapeli** [comment]
It is very likely. This only happens in three specific places, only from the default view. 

The question is whether this kind of error can be eliminated by modifying the dosbox-x code, or only by patching the game code?

**Comment by johnnovak** [comment]
I was able to reproduce the crash on real hardware using @kcgen's savegame from [this post](https://github.com/joncampbell123/dosbox-x/issues/3939#issuecomment-1385556979).

I used SVGA resolution, and the game consistently crashes when completing the first lap. There's some random garbage at the bottom of the screen, and the last note played by the AWE32 synth keeps looping indefinitely. Classic DOS crash 😎🤘🏻 

Hardware used:

- Pentium MMX 233 (underclocked to 125 MHz)
- Diamond Stealth 3D 2000 Pro (S3 ViRGE/DX)
- Sound Blaster AWE32

The native AWE32 support sounds quite nice in this game, btw 😎

![screamer-crash](https://user-images.githubusercontent.com/698770/213995766-c7492a68-18c5-484a-b3df-44a515b3402c.jpg)
